qtfred/src/ui/dialogs/ShipEditor/ShipCustomWarpDialog.cpp

```diff
@@ -0,0 +1,212 @@
+			return;
+		}
+		if (button == DialogButton::No) {
+			_model->reject();
+		}
+	}
+	QDialog::closeEvent(e);
+}
+void ShipCustomWarpDialog::updateUI(const bool firstrun)
+{
+	util::SignalBlockers blockers(this);
+	if (firstrun) {
+		for (int i = 0; i < Num_warp_types; ++i) {
+			SCP_string name = Warp_types[i];
+			if (name == "Hyperspace") {
+				name = "Star Wars";
+			}
+			ui->comboBoxType->addItem(name.c_str());
+		}
+		for (const auto& fi : Fireball_info) {
+			ui->comboBoxType->addItem(fi.unique_id);
+		}
+		ui->comboBoxType->setCurrentIndex(_model->getType());
+
+		ui->lineEditStartSound->setText(_model->getStartSound().c_str());
+		ui->lineEditEndSound->setText(_model->getEndSound().c_str());
+		if (!_model->departMode()) {
+			ui->doubleSpinBoxEngage->setEnabled(false);
+			ui->labelEngageTime->setEnabled(false);
+		} else {
+			ui->doubleSpinBoxEngage->setValue(_model->getEngageTime());
+		}
+
+		ui->doubleSpinBoxSpeed->setValue(_model->getSpeed());
+		ui->doubleSpinBoxTime->setValue(_model->getTime());
+
+		if (_model->departMode()) {
+			ui->labelExponent->setText(tr("Deceleration Exponent:"));
+		} else {
+			ui->labelExponent->setText(tr("Acceleration Exponent:"));
+		}
+		ui->doubleSpinBoxExponent->setValue(_model->getExponent());
+		ui->doubleSpinBoxRadius->setValue(_model->getRadius());
+		ui->lineEditAnim->setText(_model->getAnim().c_str());
+		if (_model->getSupercap()) {
+			ui->checkBoxSupercap->setChecked(true);
+		}
+		if (!_model->isPlayer() || !_model->departMode()) {
+			ui->labelPlayerSpeed->setEnabled(false);
+			ui->doubleSpinBoxPlayerSpeed->setEnabled(false);
+		} else {
+			ui->doubleSpinBoxPlayerSpeed->setValue(_model->getPlayerSpeed());
+		}
+	}
+
+	if (ui->comboBoxType->itemText(ui->comboBoxType->currentIndex()) == "Homeworld") {
+		ui->lineEditAnim->setEnabled(true);
+		ui->labelAnim->setEnabled(true);
+	} else {
+		ui->lineEditAnim->setEnabled(false);
+		ui->labelAnim->setEnabled(false);
+	}
+
+	if (ui->comboBoxType->itemText(ui->comboBoxType->currentIndex()) == "Star Wars") {
+		ui->doubleSpinBoxExponent->setEnabled(true);
+		ui->labelExponent->setEnabled(true);
+	} else {
+		ui->doubleSpinBoxExponent->setEnabled(false);
+		ui->labelExponent->setEnabled(false);
+	}
+}
+void ShipCustomWarpDialog::rejectHandler()
+{
+	if (_model->query_modified()) {
+		auto button = _viewport->dialogProvider->showButtonDialog(DialogType::Question,
+			"Changes detected",
+			"Do you want to keep your changes?",
+			{DialogButton::Yes, DialogButton::No, DialogButton::Cancel});
+
+		if (button == DialogButton::Cancel) {
+			return;
+		}
+
+		if (button == DialogButton::Yes) {
+			accept();
+			return;
+		}
+		if (button == DialogButton::No) {
+			_model->reject();
+			QDialog::reject();
+		}
+	} else {
+		_model->reject();
+		QDialog::reject();
+	}
+}
+void ShipCustomWarpDialog::startSoundChanged()
+{
+	// String wrangling reqired in order to avoid crashes when directly converting from Qstring to std::string on some
```

Comment:
The string wrangling should definitely be in a utility function.  If you want to do that in a different PR that's fine; but if string wrangling is necessary, then there should be qtFRED-wide utility functions for converting from Qstring to std::string and vice versa.

Reply:
Agreed, but that should probably be done in a separate PR as there are many places in existing dialogs it should be put.